File: Java/Eta/com/thomsonreuters/upa/valueadd/reactor/WlService.java
package com.thomsonreuters.upa.valueadd.reactor;

import java.util.LinkedList;

import com.thomsonreuters.upa.valueadd.common.VaNode;
import com.thomsonreuters.upa.valueadd.domainrep.rdm.directory.DirectoryMsgFactory;
import com.thomsonreuters.upa.valueadd.domainrep.rdm.directory.Service;

/* Watchlist service that contains necessary information regarding the service. */
public class WlService extends VaNode
{
    Service _rdmService = DirectoryMsgFactory.createService();
    long _numOutstandingRequests; // number of outstanding requests for service
    
    /* waiting request list to handle cases where request could not be submitted with
     * directory stream up but the open window wasn't open */
    LinkedList<WlRequest> _waitingRequestList = new LinkedList<WlRequest>();
    
    /* watchlist streams associated with this service */
    LinkedList<WlStream> _streamList = new LinkedList<WlStream>();    
    
    /* Returns the RDM service. Use to get or set RDM service. */
    Service rdmService()
    {
        return _rdmService;
    }

    /* Returns the number of outstanding requests. */
    long numOutstandingRequests()
    {
        return _numOutstandingRequests;
    }

    /* Sets the number of outstanding requests. */
    void numOutstandingRequests(long numOutstandingRequests)
    {
        _numOutstandingRequests = numOutstandingRequests;
    }
    
    /* Returns the waiting request list. Handle cases where request could not be
     * submitted with directory stream up but the open window wasn't open. */
    LinkedList<WlRequest> waitingRequestList()
    {
        return _waitingRequestList;
    }
    
    /* Returns the list of streams associated with this service. */
    LinkedList<WlStream> streamList()
    {
        return _streamList;
    }

    /* Clears the object for re-use. */
    void clear()
    {
        _numOutstandingRequests = 0;
        WlRequest wlRequest = null;
        while ((wlRequest = _waitingRequestList.poll()) != null)
        {
            wlRequest.returnToPool();
        }
        WlStream wlStream = null;
        while ((wlStream = _streamList.poll()) != null)
        {
            wlStream.returnToPool();
        }
    }    
}

File: CSharp/Eta/Src/ValueAdd/Reactor/WlService.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2023-2024 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

using LSEG.Eta.ValueAdd.Common;
using LSEG.Eta.ValueAdd.Rdm;
using Buffer = LSEG.Eta.Codec.Buffer;

namespace LSEG.Eta.ValueAdd.Reactor
{
    internal sealed class WlService
    {
        public Service? RdmService { get; internal set; }
        public long NumOutstandingRequests { get; internal set; }

        /// <summary>
        /// Gets or sets the waiting request list to handle cases where request could not be submitted
        /// with directory stream up but the open windows wasn't open
        /// </summary>
        public LinkedList<WlItemRequest> WaitingRequestList { get; set; } = new ();

        public WlStreamIdList StreamIdDlList { get; set; } = new();

        /// <summary>
        /// Dictionary of item groups, by group ID.
        /// </summary>
        public Dictionary<Buffer, WlItemGroup> ItemGroupDict = new (50);
    }
}
